### DISCUSSION TITLE
Optimise black box Simulink Model

### DISCUSSION THREAD

**Discussion by gavilomon** [casadi/casadi#3586] [Q&A]
Hello,I currently have a
Simulink model
(protected) so a black box really. Would it be possible to employ any autodiff scheme or casadi block/ tool to optimise it using a direct collocation method with Ipopt for instance? As far as I understand I would just need the derivatives of the objective fcn and model equations/ constraints.

Thank you for your attention.

Kind regards,
J.

**Comment by brunoml5** [comment]
Hi!

Probably you can. Take a look at the [blog posts](https://web.casadi.org/blog/).

**Reply by gavilomon** [reply]
Hi Bruno! Thanks for the answer. Yes I already took a look at them. It is the xdot representing the model equations the bit I am asking about. So it would be a bit like that example but making the casadi block agnostic of the model equations. The "xdot" there should be extracted from the black box model (or the derivatives of such model).

Hopefully that makes sense.

J

**Reply by brunoml5** [reply]
From what you said, maybe you must encapsulate your black box model in a Callback Function. Then, Casadi will be able to "understand" your model and connect it with the solvers and integrators.

This [block post](https://web.casadi.org/blog/tensorflow/) shows how to use Tensorflow with Casadi. Notice that it is important for speed to define some kind of derivative. This part of the [documentation](https://web.casadi.org/docs/#subclassing-callback) also explain the Callback functionality. See if this solves your issue.
